File: MacTemplet/Cagtegory/NSMenuItem+Helper.m

//
//  NSMenuItem+Helper.m
//  MacTemplet
//
//  Created by Bin Shang on 2019/6/20.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

#import "NSMenuItem+Helper.h"
#import <objc/runtime.h>

@implementation NSMenuItem (Helper)

- (void)addActionHandler:(void(^)(NSMenuItem *menuItem))handler{
    self.target = self;
    self.action = @selector(p_handleAction:);

    objc_setAssociatedObject(self, _cmd, handler, OBJC_ASSOCIATION_COPY_NONATOMIC);
}

- (void)p_handleAction:(NSMenuItem *)menuItem{
    void(^block)(NSMenuItem *menuItem) = objc_getAssociatedObject(self, @selector(addActionHandler:));
    if (block) block(menuItem);
}

- (NSMenuItem *)initWithTitle:(NSString *)string keyEquivalent:(NSString *)charCode handler:(void(^)(NSMenuItem *menuItem))handler{
    NSMenuItem * item = [[NSMenuItem alloc]initWithTitle:string action:nil keyEquivalent:charCode];
    [item addActionHandler:handler];
    return item;
}

@end

File: MacTemplet/MacCagtegory/NSMenuItem+Helper.swift
//
//  NSMenuItem+Helper.swift
//  MacTemplet
//
//  Created by Bin Shang on 2019/11/22.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

import Cocoa

@objc extension NSMenuItem {

    static public func create(title: String, keyEquivalent charCode: String, handler: @escaping (NSMenuItem) -> Void) -> NSMenuItem {
        let menuItem = NSMenuItem(title: title, action: nil, keyEquivalent: charCode)
        menuItem.addAction(handler)
        return menuItem;
    }
    /// 闭包回调
    public func addAction(_ handler: @escaping (NSMenuItem) -> Void) {
        objc_setAssociatedObject(self, UnsafeRawPointer(bitPattern: self.hashValue)!, handler, .OBJC_ASSOCIATION_COPY_NONATOMIC);
        target = self;
        action = #selector(p_invoke(_:));
    }
    
    private func p_invoke(_ sender: NSMenuItem) {
        let handler = objc_getAssociatedObject(self, UnsafeRawPointer(bitPattern: self.hashValue)!) as! ((NSMenuItem) -> Void)
        handler(sender);
    }
}
